### PR TITLE
aws cloudformation deletion policy

### PR DESCRIPTION
**Pull request by muzzamilinovaqo**

### CODE REVIEW

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'CloudFormation Deletion Policy in Use',
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:5]:
```suggestion
    title: 'CloudFormation Deletion Policy In Use',
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    category: 'CloudFormation',
+    domain: 'Application Integration',
+    severity: 'Medium',
+    description: 'Ensures that a deletion policy is used for Amazon CloudFormation stacks.',
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:9]:
```suggestion
    description: 'Ensures that deletion policy is used for Amazon CloudFormation stacks.',
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    more_info: 'AWS Cloudformation stacks should have a deletion policy, implemented with the DeletionPolicy attribute in order preserve or backup AWS resources when the stacks are deleted.',
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:10]: make it more descriptive

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    link: 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html',
+    recommended_action: 'Enable Deletion Policy attribute in the AWS Cloudformation stack.',
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:12]:
```suggestion
    recommended_action: 'Enable deletion policy attribute in the AWS Cloudformation stack.',
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                return rcb();
+            }
+
+            if (!listStacks.data.length) {
+                helpers.addResult(results, 0, 'No CloudFormation stacks found', region);
+                return rcb();
+            }
+
+            for (var stack of listStacks.data) {
+                if (!stack.StackId) return;
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:38]:
```suggestion
                if (!stack.StackId) return;
                
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                var resource = stack.StackName;
+                var templates = helpers.addSource(cache, source,
+                    ['cloudformation', 'getTemplate', region, resource]);
+
+                if (templates && templates.data && templates.data.TemplateBody) {
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:43]: add the checks if (templates.err || !templates || !templates.data) {
helpers.addResult(results, 3,
                        `Unable to query for CloudFormation stack template: ${helpers.addError(templates)}`, region, resource);

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                    var deletionPolicy = templates.data.TemplateBody.includes('DeletionPolicy');
+                    
+                    if (deletionPolicy) {
+                        helpers.addResult(results, 0,
+                            `Deletion Policy is used for CloudFormation stack "${stack.StackName}"`,
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:48]:
```suggestion
                            "Deletion Policy is used for CloudFormation stack ",
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                            region, resource);
+                    } else {
+                        helpers.addResult(results, 2, `Deletion Policy is not used for CloudFormation stack "${stack.StackName}"`,
```
Comment (fatima99s) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:51]:
```suggestion
                        helpers.addResult(results, 2, "Deletion Policy is not used for CloudFormation stack ",
```

**Review by fatima99s — CHANGES_REQUESTED** [review]

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'CloudFormation Deletion Policy In Use',
+    category: 'CloudFormation',
+    domain: 'Application Integration',
+    severity: 'Medium',
+    description: 'Ensures that deletion policy is used for Amazon CloudFormation stacks.',
+    more_info: 'CloudFormation deletes the resource and all its content during stack deletion. With the DeletionPolicy attribute you can preserve, backup a resource when its stack is deleted. A DeletionPolicy attribute can be applied to each cloudformation stack.',
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:10]:
```suggestion
    more_info: 'DeletionPolicy attribute allows to preserve and backup a resource when its stack is deleted. By default, AWS CloudFormation deletes the resource and all its content if a resource has no DeletionPolicy attribute in a template.',
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'CloudFormation Deletion Policy In Use',
+    category: 'CloudFormation',
+    domain: 'Application Integration',
+    severity: 'Medium',
+    description: 'Ensures that deletion policy is used for Amazon CloudFormation stacks.',
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:9]:
```suggestion
    description: 'Ensures that deletion policy is used for Amazon CloudFormation stacks.',
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    more_info: 'CloudFormation deletes the resource and all its content during stack deletion. With the DeletionPolicy attribute you can preserve, backup a resource when its stack is deleted. A DeletionPolicy attribute can be applied to each cloudformation stack.',
+    link: 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html',
+    recommended_action: 'Enable deletion policy attribute in the AWS Cloudformation stack.',
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:12]:
```suggestion
    recommended_action: 'Add DeletionPolicy attribute in the AWS Cloudformation stack template.',
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                return rcb();
+            }
+
+            if (!listStacks.data.length) {
+                helpers.addResult(results, 0, 'No CloudFormation stacks found', region);
+                return rcb();
+            }
+
+            for (var stack of listStacks.data) {
+                if (!stack.StackId) return;
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:38]:
```suggestion
                if (!stack.StackName) continue;
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                
+                var resource = stack.StackName;
+                var templates = helpers.addSource(cache, source,
+                    ['cloudformation', 'getTemplate', region, resource]);
+                
+                if (!templates || templates.err || !templates.data) {
+                    helpers.addResult(results, 3, `Unable to query CloudFormation stack template: ${helpers.addError(templates)}`, region, resource);
+                    return rcb();
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:46]: should not use return rcb() inside the for loop

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+                }
+
+                if (templates && templates.data && templates.data.TemplateBody) {
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:49]:
```suggestion
                if (templates.data.TemplateBody) {
```

File: plugins/aws/cloudformation/cloudformationDeletionPolicy.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+                    var deletionPolicy = templates.data.TemplateBody.includes('DeletionPolicy');
+                    
+                    if (deletionPolicy) {
+                        helpers.addResult(results, 0,
+                            'Deletion Policy is used for CloudFormation stack ',
```
Comment (alphadev4) [inline plugins/aws/cloudformation/cloudformationDeletionPolicy.js:54]:
```suggestion
                            'Deletion Policy is used for CloudFormation stack',
```

**Review by alphadev4 — CHANGES_REQUESTED** [review]

**Review by alphadev4 — APPROVED** [review]

**Pull request merged**